kite-data/kite-data-mapreduce/src/main/java/org/kitesdk/data/mapreduce/DatasetKeyInputFormat.java

```diff
@@ -134,9 +145,22 @@ public ConfigBuilder readFrom(String uri) {
      * @return this for method chaining
      */
     public <E> ConfigBuilder withType(Class<E> type) {
+      String readerSchema = conf.get(KITE_READER_SCHEMA);
+      Preconditions.checkArgument(DataModelUtil.isGeneric(type) || readerSchema == null,
+        "Can't configure a type when a reader schema is already set: {}", readerSchema);
       conf.setClass(KITE_TYPE, type, type);
       return this;
     }
+
+    public ConfigBuilder withSchema(Schema readerSchema) {
+      Class<?> type = conf.getClass(KITE_TYPE, null);
+      Preconditions.checkArgument(type == null,
```

Comment:
It looks like `withType` allows setting a generic type. In that case, it would make sense to call both withType and withSchema.